File: pubmed_trend.pl
#! /usr/bin/perl -w
#
# sex_diff_pain.pl
#
# Created by David Ruau on 2011-02-17.  
# Department of Pediatrics/Div. System Medicine Stanford University.
# 
##################### USAGE #########################
# 
# Query PubMed with Eutils tools
# 
#####################################################
use Bio::TGen::EUtils;

use strict;

my $queryString = $ARGV[0];

## query info
my $eu = Bio::TGen::EUtils->new( 'tool'  => 'pubmed_trend.pl',
                                 'email' => 'REPLACE_ME@gmail.com' );

## EFetch
my $query = $eu->esearch( db   => 'pubmed',
                          term => $queryString,
                          usehistory => 'n' );

$query->write_raw( file => 'tempfile.xml' );

if (-z 'tempfile.xml') {
  # one more time
  my $query = $eu->esearch( db   => 'pubmed',
                            term => $queryString,
                            usehistory => 'n' );

  $query->write_raw( file => 'tempfile.xml' );
  if (-z 'tempfile.xml') {
    open (FILE, '>', 'tempfile.xml') or die 'Could not open file, $!';
    
    print FILE "<begin>hello world</begin>";
    close (FILE);
  }
}

File: pubmed_trend.r
#
# sex_diff_pain.r
#
# Created by David Ruau on 2011-02-17.  
# Department of Pediatrics/Div. System Medicine Stanford University.
# 
##################### USAGE #########################
# 
# source('pubmed_trend.r')
# sex.pub <- pubmed_trend(search.str = 'Sex+Characteristics[mh] AND Pain[mh]', year.span=1970:2011)
# 
#####################################################

pubmed_trend <- function(search.str = 'Sex+Characteristics[mh] AND Pain[mh]', year.span=1970:2011) {
  require(XML)
  require(RCurl)
  
  results <- NULL
  tmpf <- "./tempfile.xml"
  ## clean before
  system(paste("rm", tmpf))

  for(i in year.span){
    queryString <- paste(search.str, ' AND ', i, '[dp]', sep="")
    print(paste('queryString:', queryString))
    sysString <- paste('./pubmed_trend.pl "', queryString,'"', sep="")
    system(sysString)

    xml <- xmlTreeParse(tmpf, useInternalNodes=TRUE)
    pubTerm <- as.numeric(xmlValue(getNodeSet(xml, "//Count")[[1]]))
    print(paste("#______num pub for",i,":",pubTerm))
    rm(xml)
    results <- append(results, pubTerm)
    ## avoid being kicked out!
    Sys.sleep(1)
  }
  names(results) <- year.span
  ## clean after
  system(paste("rm", tmpf))
  
  return(results)
}